test/deflateparams.c

```diff
@@ -0,0 +1,254 @@
+    c_stream.avail_in = (uint32_t)len;
+    c_stream.next_out = compr;
+    c_stream.avail_out = (uint32_t)compr_bound;
+
+    err = PREFIX(deflate)(&c_stream, Z_FINISH);
+    if (err != Z_STREAM_END)
+        CHECK_ERR(err, "deflate");
+
+    err = PREFIX(deflateEnd)(&c_stream);
+    CHECK_ERR(err, "deflateEnd");
+
+    fwrite(compr, 1, c_stream.total_out, stdout);
+    free(compr);
+}
+
+/* ===========================================================================
+ * Usage:  deflateparam [-f] [-h] [-R] [-F] [-m 1 to 8] [-w #] [-1 to -9] [input file]
+ *   -f : compress with Z_FILTERED
+ *   -h : compress with Z_HUFFMAN_ONLY
+ *   -R : compress with Z_RLE
+ *   -F : compress with Z_FIXED
+ *   -m : memory level 1-8
+ *   -w : window bits
+ *   -s : small buffer
+ *   -l : large buffer
+ *   -0 to -9 : compression level
+ */
+
+int main(int argc, char **argv) {
+    int32_t i;
+    int32_t mem_level = DEF_MEM_LEVEL;
+    int32_t window_bits = MAX_WBITS;
+    int32_t strategy = Z_DEFAULT_STRATEGY;
+    int32_t level = 6;
+    uint8_t small_buf = 0;
+    uint8_t large_buf = 0;
+    size_t len, read;
+    unsigned char *buf;
+    FILE *f;
+
+
+    for (i = 1; i < argc; i++) {
+        if ((strcmp(argv[i], "-m") == 0) && (i + 1 < argc))
+            mem_level = atoi(argv[++i]);
+        else if ((strcmp(argv[i], "-w") == 0) && (i + 1 < argc))
+            window_bits = atoi(argv[++i]);
+        else if (strcmp(argv[i], "-s") == 0)
+            small_buf = 1;
+        else if (strcmp(argv[i], "-l") == 0)
+            large_buf = 1;
+        else if (strcmp(argv[i], "-f") == 0)
+            strategy = Z_FILTERED;
+        else if (strcmp(argv[i], "-h") == 0)
+            strategy = Z_HUFFMAN_ONLY;
+        else if (strcmp(argv[i], "-R") == 0)
+            strategy = Z_RLE;
+        else if (argv[i][0] == '-' && argv[i][1] >= '0' && argv[i][1] <= '9' && argv[i][2] == 0)
+            level = argv[i][1] - '0';
+        else
+            break;
+    }
+
+    printf("deflateParams: level %d memLevel %d windowBits %d strategy %d\n", level, mem_level, 
+        window_bits, strategy);
+
+    if (i == argc) {
+        fprintf(stderr, "No input file specified\n");
+        exit(1);
+    }
+
+    f = fopen(argv[i], "rb+");
+    if (f == NULL) {
+        /* Failed to open this file: it may be a directory. */
+        fprintf(stderr, "Failed to open file: %s\n", argv[i]);
+        exit(1);
+    }
+
+    fseek(f, 0, SEEK_END);
+    len = ftell(f);
+    fseek(f, 0, SEEK_SET);
+
+    buf = (unsigned char *)malloc(len);
+    read = fread(buf, 1, len, f);
+    assert(read == len);
+
+    if (small_buf) {
+        deflate_params_small(buf, len, level, window_bits, mem_level, strategy);
+    }
+    else if (large_buf) {
+        deflate_params_large(buf, len, level, window_bits, mem_level, strategy);
+    }
+    else {
+        deflate_params(buf, len, level, window_bits, mem_level, strategy);
+    }
+
+    free(buf);
+    fclose(f);
```

Comment:
Maybe these lines should be broken as separate function with file name as first parameter.

Reply:
I was going toward something more like what minigzip has - in that it can work with file in/out or std in/out. But right now for the test all it needs is file in/out.

Reply:
I've seen applications where "-" as filename means standard input... 

Long methods are another issue specific with C as all variables need to be declared at the top and compiler needs to know how many variables can stay in registers. For that it needs to count how far from top any variable is modified or read, as those that are only used close to top can be safely spilled to stack after they are no longer needed and the register can be reused for another variable. More there is calls to other functions, more pressure there is for storing variables in registers instead of stack (memory).